src/ssh/allowed_signer.rs

```diff
@@ -0,0 +1,456 @@
+                },
+                Err(_) => {
+                    return Err(Error::ParsingError);
+                },
+            };
+            allowed_signers.push(allowed_signer);
+        }
+
+        Ok(AllowedSigners(allowed_signers))
+    }
+}
+
+/// A type used to split the allowed signer segments, abstracting out the handling of double quotes.
+/// The splitter is highly aware of the allowed_signer format and will catch certain invalid
+/// formats.
+///
+/// For example: "principals   option1=\"value1   value2\" option2 option3=value kt key_data" is split into
+/// ["principals", "option1=\"value1   value2\"", "option2", "option3=value", "kt", "key_data"]
+struct AllowedSignerSplitter {
+    /// A buffer of remaining tokens in reverse order.
+    buffer: Vec<String>,
+}
+
+impl AllowedSignerSplitter {
+    /// Split the string by delimiters but keep the delimiters.
+    pub(in self) fn new(s: &str) -> Self {
+        let mut buffer = Vec::new();
+        let mut last = 0;
+
+        for (index, matched) in s.match_indices([' ', '"', '#']) {
+            // Push the new text before the delimiter
+            if last != index {
+                buffer.push(s[last..index].to_owned());
+            }
+            // Push the delimiter
+            buffer.push(matched.to_owned());
+            last = index + matched.len();
+        }
+
+        // Push the remaining text
+        if last < s.len() {
+            buffer.push(s[last..].to_owned());
+        }
+
+        // We parse from left to right so reversing allow us to use Vec's last() and pop()
+        buffer.reverse();
+
+        Self { buffer }
+    }
+
+    pub(in self) fn is_empty_after_trim(&mut self) -> bool {
+        self.trim();
+        return self.buffer.is_empty();
+    }
+
+    /// Get the next part that is not an option (principals, key)
+    /// If opening_quotes_allowed is set to false, we reject the next token if it starts with ".
+    pub(in self) fn next(&mut self, opening_quotes_allowed: bool) -> Result<Option<String>> {
+        if self.is_empty_after_trim() {
+            return Ok(None);
+        }
+
+        // If the next token starts with a double quote, then the closing double quote is also
+        // the end of the token
+        if self.buffer[0] == "\"" {
+            if opening_quotes_allowed {
+                return self.split_quote().map(|v| Some(v));
+            } else {
+                return Err(Error::InvalidAllowedSigner(AllowedSignerParsingError::InvalidQuotes));
+            }
+        }
+
+        // If the next token doesn't start with a double quote, the token can represent an option.
+        // Only an option token can contain double quotes in the middle (e.g. a="b c").
+        // If we don't see any double quote in the token, we greedily parse the token until the
+        // next whitespace.
+        let mut s = String::new();
+        while !self.buffer.is_empty()
```

Comment:
You might be able to get rid of the the `last().unwrap()` by using a `while let` here

Reply:
I've used `while let` where I can. I also tried to use `if let` to remove the unwrap but that made things look a little uglier since `if let _ && cond` is still unstable.